Add Tutorial section (#60)
empa-scientific-it/how-to-git: src/content/docs/tutorial/2-tutorial.mdx

</Aside>

<Steps>

1. **Go to the Issues tab.**
   Click on :radio_button: `Issues` in the top navigation of the repository.

2. **Create a new issue.**
   Click the green `New issue` button in the top-right corner.

3. **Write a clear title.**
   The title should describe *what* you plan to do.

4. **Add a description.**
   In the text area below, briefly explain *why* this change is needed.
   You can use Markdown syntax to format text.

5. **Assign yourself.**
   On the right-hand side, click `Assignees` and select your own GitHub username.
   Adding an :term[assignee] signals to the team that there is a person responsible for this task.

6. **Add a label.**
   Click `Labels` and select a relevant :term[label].
   This helps with filtering and prioritizing issues.

7. **Submit the issue.**
   Click the green `Create` button.
   You will be taken to the issue page.
   Note the issue number (e.g., `#7`) shown next to the title - you will need it later.

</Steps>

<Aside type="tip">
Keep the issue page open in a browser tab.
In the next steps you will create a dedicated branch for this issue and you will link your pull request back to it.
</Aside>

---

> ## Part 3: Create a Branch

Reviewer: Add a "Create a Fork" section; otherwise, people won't be able to create branches.

   On the right-hand side panel, scroll down until you see the `Development` section.
   Click `Create a branch`.

3. **Choose a branch name.**
   GitHub suggests a name based on the issue number and title, for example `7-issue-name`.
   It is best to keep this name as it clearly links the branch to the issue.
   You can of course shorten it if it is too long, but remember to keep the issue number at the start of the branch name.

4. **Leave the other options at their defaults.**
   - Repository destination: the shared repository
   - Branch source: `main`

5. **Click the green `Create branch` button.**
   GitHub may show a dialog with `git` commands to checkout locally.
   You can ignore this since for this tutorial we will only work in the browser.

6. **Switch to your new branch.**
   Go to the repository home page and click on the branch dropdown.
   Select your new branch from the list and the page will reload showing your branch.
   Notice the branch name is now shown in the dropdown, instead of `main`.

</Steps>

---

## Part 4: Make a Change and Commit It

Now that you are on your own branch, you can safely make changes without affecting anyone else's work.

<Steps>

> 1. **Navigate to the correct folder.**

Reviewer 2: this isn't clear maybe something like
```suggestion
1. **Navigate to the folder where you want to add a text file.**
```

2. **Create a new file.**
   Click the `Add file` button at the top-right area of the file list and select `Create new file`.

3. **Name your file.**
   In the filename field at the top, type the path to your file, followed by the file's name.

4. **Write your content.**
   In the editor, add a few lines.
   This does not need to be perfect now, we will come back to edit the content later.

5. **Commit your change.**
   Click the `Commit changes...` button in the top-right.
   A dialog box will appear.

6. **Write a meaningful commit message.**
   Replace the default message with something descriptive.
   Optionally add an extended description if there is more context to provide.

7. **Make sure the correct branch is selected.**
   The dialog should show "Commit directly to the `7-your-branch` branch".
   Do **not** select `main`.

8. **Finish the commit.**
   Click the `Commit changes` button.
   Your file is now saved and tracked in your branch.

</Steps>

<Aside type="tip">

> Remember, every commit is a snapshot that you can return to at any time.

Reviewer: Technically, commit is not a snapshot, is literally a difference (that's why they are so small). But I am not sure how precise we want to be here, all in all, what you say is kind of true in the sense that clicking on a commit brings you to the the state of repo where all changes up to the selected commit are applied.

I hope this is understandable, cannot formulate it better, sorry :).

</Aside>

---

## Part 5: Open a Pull Request

A Pull Request, also known as PR, is a formal request to merge your branch into `main`.
It is also the space where collaborators review your changes and give feedback.

<Steps>

1. **Start a new pull request.**
   Go to the `Pull requests` tab.
   Click the green `New pull request` button.

2. **Select your branch.**

>    - `Base branch`: `main` (this is where your changes will be merged into)

Reviewer 2: here specify what the origin is, just more prcise

Author: not needed since we will not be working with forks at the end

   - `Compare branch`: your branch (e.g., `7-your-branch`)

   After selecting, GitHub will show you a diff — the exact changes you are proposing.

3. **Start the pull request creation.**
   Click the green `Create pull request` button.

4. **Write a title and description.**
   - The PR title should be concise and descriptive.
   - In the description, explain what your PR does and why.
   Then link it to your issue by adding a line like:
   ```
   Closes #7
   ```
   where you replace `7` with your actual issue number.
   When the PR gets merged, GitHub will automatically close the linked issue.

> 5. **Request a reviewer.**

Reviewer 2: probably not possible, to be checked

   On the right-hand side, click `Reviewers` and select a classmate to review your PR.
   Ask the classmate sitting next to you - you will review each other's PRs.

6. **Finish the pull request creation.**
   Click the green `Create pull request` button.
   Your PR is now open and your reviewer will be notified.

</Steps>

<Aside type="note" title="Draft pull requests">
If you are not finished yet with your work, but still want to share your progress, you can open a **Draft Pull Request** by clicking the dropdown arrow next to the button for pull request creation and selecting `Create draft pull request`.
A draft PR cannot be merged until it is marked as ready for review.
</Aside>

---

## Part 6: Review a Pull Request

> While you wait for your reviewer to look at your PR, it is your turn to review the PR that your classmate has assigned to you.

Reviewer 2: not necessarily assigned to you

Reviewing is an essential skill - it helps catch problems early and improves the quality of the shared work.

<Steps>

1. **Find the PR assigned to you.**
   Click on the tab `Pull requests` in the top navigation of the repository.
   Look for the PR where your classmate has requested your review.

>    You may also have received a notification by email or on GitHub.

Reviewer 2: doesn't happen

> 2. **Read the description.**

Reviewer 2: what is the description? you mean the conversation?

2. **Check the review status.**
   If your reviewer requested changes, address the feedback first:
   - To accept an inline suggestion, click the green `Apply suggestion` button directly on the comment.
   - To make other changes, go to your branch, edit the file, and commit again. The PR updates automatically.
   - Once you have addressed everything, click the green `Re-request review` button to notify your reviewer.

3. **Resolve conversations.**
   After addressing feedback, mark each comment thread as `Resolved`.

4. **Merge the pull request.**
   When the PR is approved, click the green `Merge pull request` button.
   Confirm by clicking `Confirm merge`.

5. **Delete the branch.**
   After merging, GitHub shows a button to `Delete branch`.
   Click it, then the branch has served its purpose and can be deleted.

6. **Check that your issue is closed.**
   Go to the `Issues` tab.
   Because you wrote `Closes #7` in the PR description, the issue should now be automatically closed.

7. **View the network graph.**
   Go to the `Insights > Network` tab again.
   You will now see your branch joining the `main` line - a visual record of your contribution.

> 

Reviewer 2: approval from maintainers is required

</Steps>

<Aside type="tip">
Deleting the merged branch keeps the repository tidy.
The full history of the commits is preserved in `main`, so nothing is lost.
</Aside>

---

## Part 8: Resolve a Conflict

A **merge conflict** occurs when two branches have changed the same part of the same file in different ways.
Git cannot decide which version to keep automatically so it needs a human to choose.

<Aside type="note" title="Setup">
We will now introduce a conflict scenario.
Two participants will have edited the **same lines** in the same file on **different branches**.
When one branch is merged first, the second branch will have a conflict that must be resolved before it can also be merged.
</Aside>

### How a conflict happens

Imagine two people both edit `README.md` on their respective branches:
- **Branch A** changes the project title to `"Empa Git Workshop"`.
- **Branch B** changes the same line to `"Empa GitHub Tutorial"`.

When Branch A is merged first, Branch B now conflicts with `main`.

> ### Resolve the conflict in the GitHub web interface

Reviewer 2: to do this, add a contributors file to the root folder, where everyone should add their name, conflicts arrive automatically

<Steps>

1. **Navigate to the folder where you want to add a text file.**

2. **Create a new file.**
   Click the `Add file` button at the top-right area of the file list and select `Create new file`.

3. **Name your file.**
   In the filename field at the top, type the path to your file, followed by the file's name.

4. **Write your content.**
   In the editor, add a few lines.
   This does not need to be perfect now, we will come back to edit the content later.

5. **Commit your change.**
   Click the `Commit changes...` button in the top-right.
   A dialog box will appear.

6. **Write a meaningful commit message.**
   Replace the default message with something descriptive.
   Optionally add an extended description if there is more context to provide.

7. **Make sure the correct branch is selected.**
   The dialog should show "Commit directly to the `7-your-branch` branch".
   Do **not** select `main`.

8. **Finish the commit.**
   Click the `Commit changes` button.
   Your file is now saved and tracked in your branch.

</Steps>

<Aside type="tip">
Remember, every commit is also a snapshot that you can return to at any time.

> Clicking on a commit brings you to the state of the repository where all changes up to the selected commit are applied.

Reviewer: Still not a snapshot; maybe something like this? It is vaguer but more accurate as well.

Up to you to decide; if you still prefer your version, keep it.

```suggestion
Remember, clicking on a commit takes you to the state of the repository where all changes, including that commit, are applied.
```

   - To accept an inline suggestion, click the green `Apply suggestion` button directly on the comment.
   - To make other changes, go to your branch, edit the file, and commit again. The PR updates automatically.
   - Once you have addressed everything, click the green `Re-request review` button to notify your reviewer.

3. **Resolve conversations.**
   After addressing feedback, mark each comment thread as `Resolved`.

4. **Merge the pull request.**
   When the PR is approved, click the green `Merge pull request` button.
   Confirm by clicking `Confirm merge`.

5. **Delete the branch.**
   After merging, GitHub shows a button to `Delete branch`.
   Click it, then the branch has served its purpose and can be deleted.

6. **Check that your issue is closed.**
   Go to the `Issues` tab.
   Because you wrote `Closes #7` in the PR description, the issue should now be automatically closed.

7. **View the network graph.**
   Go to the `Insights > Network` tab again.
   You will now see your branch joining the `main` line - a visual record of your contribution.

</Steps>

<Aside type="tip">
Deleting the merged branch keeps the repository tidy.

> The full history of the commits is preserved in `main`, so nothing is lost.

Reviewer: This is not fully true, especially if the PR was squash-merged.

```suggestion
The full history of the commits can be looked up later by re-visiting the merged PR.
If needed, the branch corresponding to the PR can be restored from the GitHub interface.
```
